Reliable delivery in Typed, #20984

patriknw opened the pull request:
## Approach

I'm taking a rather different approach than in classic `AtLeastOnceDelivery` because I would like:

* support plain flow control, with a work pulling approach
* be possible to use with or without persistence (without it may loose messages if producer node crashes)
* detect lost messages on the consumer side and let that drive resends, instead of aggressively resending from producer side
* deliver messages in order and deduplicate resent messages
* have an efficient protocol for acknowledgments over the network (not ack each message), but still have a simple one-by-one protocol for the end user

I have thought of 3 separate usage cases:

1. Point-to-point between one producer and one consumer, this is also the building block of other usages.
1. Work pulling pattern, where workers register to one work producer master. Workers are equal and a message can be routed to any of the workers that has demand.
1. Sharding. Routing to the destination is based on the entityId. Many producers can send to the same destination entity.

Those use cases are illustrated in the tests in `ReliableDeliverySpec`, `WorkPullingSpec`, `ReliableDeliveryShardingSpec`.

## Point-to-point

ProducerController and ConsumerController are point-to-point with single producer and single consumer.

The ScalaDoc in `ProducerController` and `ConsumerController` describes how it works and the message protocol.

![Akka docs diagrams 001](https://user-images.githubusercontent.com/336161/73140334-16bd9d80-4078-11ea-9e63-882ce870fd2c.png)

## Work Pulling

ConsumerController (CC-1) is started for worker-1, and registered to the Receptionist.

The WorkPullingProducerController receives the Listing from the Receptionist and spawns a ProducerController (PC-1) for CC-1. The PC-1/CC-1 pair is used for delivery of messages selected for worker-1.

![Akka docs diagrams 002](https://user-images.githubusercontent.com/336161/73140346-289f4080-4078-11ea-937a-7759f49bb8b5.png)

Another worker-2 is started and registered. Corresponding ProducerController (PC-2) for CC-2. The PC-1/CC-1 pair is used for delivery of messages selected for worker-1.

The WorkPullingProducerController randomly selects worker for each message (job) among the workers with demand (according to the PC/CC pair request window).

![Akka docs diagrams 003](https://user-images.githubusercontent.com/336161/73140358-3bb21080-4078-11ea-9192-b6835ee56a45.png)

## Sharding

A ShardingProducerController is started for each producer, typically one per node. Messages to any entity can be sent via this ShardingProducerController. For the first message for an entity it spawns a ProducerController (PC-entity-1), which is used for delivery of messages to that specific entity.

Sharding creates the ShardingConsumerController for the first message to that entity, which spawns the application's entity Behavior.

A ConsumerController (CC-a) is also spawned and is working in pair with the ProducerController (PC-entity-1).

![Akka docs diagrams 004](https://user-images.githubusercontent.com/336161/73140382-82a00600-4078-11ea-8cf5-c658130bc719.png)

Message for another entity means that the ShardingProducerController will spawn another ProducerController (PC-2) for that entity.

![Akka docs diagrams 005](https://user-images.githubusercontent.com/336161/73140386-90558b80-4078-11ea-9b1f-e39315b6d764.png)

There can be many producers, typically one per node. The ShardingConsumerController will spawn a new ConsumerController for each producer.

![Akka docs diagrams 006](https://user-images.githubusercontent.com/336161/73140395-a2cfc500-4078-11ea-9f64-ea7250000230.png)

## TODO:

- [x] actual public api
- [x] more around the work pulling support
- [x] more around the sharding support
- [x] real impl of the DurableProducerQueue
- [x] configuration
- [x] logging
- [x] protobuf serialization
- [x] api and reference documentation
- [x] other fixmes

Refs #20984

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,333 @@
+ * The `ProducerController` sends `RequestNext` to the actual producer, which is then allowed to send
+ * one message to the `ProducerController`. The `RequestMessage` message is defined via a factory
+ * function so that the producer can decide what type to use. The producer and `ProducerController`
+ * are supposed to be local so that these messages are fast and not lost.
```
Interesting pushing the adaption to the sending side instead of requiring a message adapter, is it worth the convenience with another pattern for it though? And if it is, should we use it in more places?

patriknw replied:
I'll take a look to see if we can use a message adapter in the P instead. That would require that P is started, creates the adapter, and then sends an initial Start message to the PC.

On the CC side I used message adapter and such start message.

patriknw replied:
I have completed this refactoring. Now the producer initiates the flow by sending a Start to the ProducerController. I think that is better.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,333 @@
+ * also a way to connect the ProducerController and ConsumerController in a dynamic way, for
+ * example when the ProducerController is replaced.
+ *
+ * When the first message is received by the `ConsumerController` it sends back the initial `Request`.
```
I read the implicit as having a deliver->confirmed roundtrip as well before it sends the initial  `Request`, am I mistaken?

patriknw replied:
first Request is sent back immediately, there is one additional Request sent back for the first 1L message after the Confirmation, but the purpose of that is to confirm that the first has been completed to cancel the special resending of the first from PC. I will probably also introduce a Ack message for other reasons, and that can be used for that latter case.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,333 @@
+ * to ignore lost messages, and then the `ProducerController` will not buffer unconfirmed messages.
+ * In that mode it provides only flow control but no reliable delivery.
+ */
+object ProducerController {
```
Is there a way for the producer to signal that it is done/reached the end of the messages it wants to send etc?

patriknw replied:
no, good idea

patriknw replied:
Changing my mind, better to have that as an application level end message. Because it needs the same delivery mechanism and the application consumer must be able to understand it. When it has been received the application consumer can tear things down.

johanandren replied:
Ok, so that means the CC stopping on consumer termination?

patriknw replied:
yes, that is implemented

on producer side I think the producer would have to wait for ack corresponding to that last message before stopping the PC

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,333 @@
+
+  final case class RequestNextParam[A](currentSeqNr: Long, confirmedSeqNr: Long, sendNextTo: ActorRef[A])
+
+  final case class MessageWithConfirmation[A](message: A, replyTo: ActorRef[Long])
```
Perhaps dont take a boxed-primitive as reply message but a wrapping type? 

So the idea is that the message is `A` but the response is the sequence number it got? How will a producer correlate those two?

patriknw replied:
The primary purpose is to use this with `ask` from the P. The reply (`seqNr: Long`) is sent back when all is confirmed for that `message: A`. Having the seqNr as the reply might not be important, could be `Done`. I exposed seqNr to both P and C so far.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,333 @@
+  final case class RegisterConsumer[A](
+      consumerController: ActorRef[ConsumerController.Command[A]],
+      replyTo: ActorRef[Done])
+      extends Command[A]
```
This is just here for testability?

patriknw replied:
no, needed for dynamic registration/replacement of CC, for example when using sharding

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,120 @@
+      hasRequested: Boolean)
+
+  // TODO Now the workers have to be registered explicitly/manually.
+  // We could support automatic registration via Receptionist, similar to how routers work.
```
Was just going to ask if you had thought about how to find the remote consumer, for this case it makes sense, one producer many consumers/workers. But for a 1:1 reliable delivery between cluster nodes that is potentially short lived I don't think the receptionist is a great fit (maybe I just made that use case up though, not sure) .

patriknw replied:
I agree, Receptionist shouldn't be used for short lived.
The point-to-point scenario currently requires some application specific way to exchange CC/PC ActorRef to connect them.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,120 @@
+          State(Map.empty, hasRequested = false))
+      }
+      .narrow
+  }
```
This is very cool, kind of pre-baking much of the distributed workers sample.

patriknw replied:
yup, was fun to write

johanandren reviewed (commented):
Through a first skim, looks very cool, but have lots of questions :)

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,308 @@
+          Behaviors.same // ignore until we receive the expected
+        }
+
+      case RetryRequest ⇒
```
it's a receiveTimeout, so retries are triggered when things are stuck

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,308 @@
+  }
+
+  // The message has been delivered to the consumer and it is now waiting for Confirmed from
+  // the consumer. New SequencedMessage from the ProducerController will be stashed.
```
No, between PC and CC it's a demand window, not one-by-one.
The Request from CC to PC defines to the upper seqNr that the CC would like, PC can continue sending until reaching that limit. New Request is sent from CC after window/2.

Between P and PC it's one-by-one because that's easier for end user to reason about and that is quick local messages. Same, one-by-one between CC and C.

hepin1989 commented:
>detect lost messages on the consumer side and let that drive resends, instead of aggressively resending from producer side

with a sequential seqNr or syncId?

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/DurableProducerState.scala:
```diff
@@ -0,0 +1,29 @@
+
+import akka.actor.typed.ActorRef
+
+object DurableProducerState {
```
I have added a first sketch of how to add the optionally durable part of the ProducerController. See https://github.com/akka/akka/pull/28155/commits/4afad2f17c1116693a7ecefb0bc8480e00f55bbd

Not tested at all yet, because first I would like feedback on the direction.

This `DurableProducerState` is supposed to be and actor with below message protocol. We provide an implementation with `EventSourcedBehavior`.

The `ProducerController` asks StoreMessageSent before sending the message, and `StoreMessageConfirmed` when confirmation is received.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,364 @@
+ * it sends [[ProducerController.Internal.Resend]] to the `ProducerController` and will ignore all messages until
+ * the expected sequence number arrives.
+ */
+object ConsumerController {
```
One question for the public api design is if we need scaladsl/javadsl for this? I can't see anything public so far that is different for Scala/Java.

It would be some dance to define (same) messages in both scaladsl and javadsl. Maybe just stick to one common api here?

package: `akka.actor.typed.delivery`?

johanandren replied:
Sounds good to share API and that use that package 👍

chbatey replied:
+1 to akka.actor.typed.delivery

hepin1989 replied:
akka.actor.typed.reliable/sync?

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,493 @@
+   * For sending confirmation message back to the producer when the message has been fully delivered, processed,
+   * and confirmed by the consumer. Typically used with `ask` from the producer.
+   */
+  final case class MessageWithConfirmation[A](message: A, replyTo: ActorRef[Long]) extends InternalCommand
```
`ActorRef[Long]` doesn't work for Java, I think. We can add some real class for that reply. Good anyway.

patriknw commented:
Updated with some important improvements to WorkPulling

patriknw commented:
the test failures are real, but I haven't been able to reproduce locally and understand the reason yet

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,308 @@
+          context.log.infoN("from producer [{}], missing [{}], received [{}]", pid, expectedSeqNr, seqNr)
+          if (resendLost) {
+            seqMsg.producer ! Resend(fromSeqNr = expectedSeqNr)
+            resending(s.copy(producer = seqMsg.producer))
```
no time/size window here before sending the `resending` message?
I more prefer to change the behavior to some kind of `observing` before that

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,308 @@
+          } else {
+            deliverTo ! Delivery(pid, seqNr, msg, context.self)
+            waitingForConfirmation(s.copy(producer = seqMsg.producer, receivedSeqNr = seqNr), first)
+          }
```
better a strategy?

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,405 @@
+          context.log.infoN("from producer [{}], ignoring [{}], waiting for [{}]", pid, seqNr, s.receivedSeqNr + 1)
+          if (seqMsg.first)
+            retryRequest(s)
+          Behaviors.same // ignore until we receive the expected
```
stashing ?

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,414 @@
+
+  private val RequestWindow = 20 // FIXME should be a param, ofc
+
+  def apply[A](resendLost: Boolean): Behavior[Command[A]] = {
```
In what scenario would you not want to resend lost, doesn't that take make it unreliable rather than reliable?

patriknw replied:
yes, then it's not reliable but still has flow control, which is useful in itself

if we think it adds too much complexity to support both we could remove that part

patriknw replied:
the main performance benefit of the flow-control-only mode is that the PC doesn't have to buffer pending messages

johanandren replied:
Ah, now I remember talking about the flow control as well previously. Thx for clarifying.

patriknw replied:
there are 3 levels
* flow control (not reliable)
* in-memory reliability
* durable reliability

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,414 @@
+      /** INTERNAL API */
+      @InternalApi private[akka] val producer: ActorRef[ProducerController.InternalCommand])
+      extends Command[A]
+  private final case object Retry extends InternalCommand
```
ok, I'll look into that - for the Request timeout I'm maybe worried about that it would mean a lot of re-scheduling which could have a performance cost when high throughput

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/internal/delivery/ConsumerControllerSpec.scala:
```diff
@@ -0,0 +1,362 @@
+      consumerController ! ConsumerController.Confirmed(3)
+      producerControllerProbe.expectMessage(ProducerController.Internal.Request(3, 20, true, true))
+
+      testKit.stop(consumerController)
```
Normally the Request is sent when half of the window has been used. The window is hardcoded to 20 at the moment, so without the timeout resending there wouldn't be one until 10 was confirmed by consumer. Both 2 and 3 are triggered by those timers.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,506 @@
+          start.producer ! RequestNext(producerId, s.currentSeqNr, s.confirmedSeqNr, msgAdapter, ctx.self)
+        active(s.copy(producer = start.producer))
+
+      case RegisterConsumer(consumerController: ActorRef[ConsumerController.Command[A]] @unchecked) =>
```
Not that I love long class names but perhaps RegisterConsumerController would be worth it here since there also is a Consumer in the problem domain.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,506 @@
+
+    def onMsg(m: A, newPendingReplies: Vector[(Long, ActorRef[Long])], ack: Boolean): Behavior[InternalCommand] = {
+      checkOnMsgRequestedState()
+      // FIXME adjust all logging, most should probably be debug
```
General thought: perhaps we should also start using trace for high-volume/per-message-debg logging now that we have access to that through SLF4J?

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,506 @@
+      confirmedSeqNr: Long,
+      requestedSeqNr: Long,
+      pendingReplies: Vector[(Long, ActorRef[Long])],
+      supportResend: Boolean,
```
This confuses me a bit, that it is both in the state of the producer, and in the Request message, whether to support resends. For example on receive `Request` half of the logic (in `onAck`) uses the flag from state, and half from the message. Would it ever change after the PC has been started?

patriknw replied:
I wanted to drive the setting from the CC. It would be more intutive if it was passed to PC in the RegisterConsumer instead of in each Request, but if I remember correctly that doesn't work for the sharding scenario, because then there is no RegisterConsumer. Then they are connected via the fist SequencedMessage since PC sends via sharding.

You are right that it doesn't change. I'll look at the logic if something can be cleaned up.

patriknw commented:
I have implemented the support for sharding, including DurableQueue for it (well, still using simulated sharding and fake DurableQueue).

It's in a good shape for another round of reviews before I continue with the final tasks, such as serialization, real sharding, real DurableQueue, cleanup logging, remaining fixmes, and documentation.

johanandren commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/internal/delivery/ReliableDeliveryShardingSpec.scala:
```diff
@@ -0,0 +1,297 @@
+    private def idle(n: Int): Behavior[Command] = {
+      Behaviors.receiveMessage {
+        case Tick                => Behaviors.same
+        case RequestNext(sendTo) => active(n + 1, sendTo)
```
I'm struggling a bit to see how I'd actually use this

Since this is pulled work, the most common use case of using sharding, like a HTTP endpoint posting one off data for an entity would not be a good fit for reliable delivery since it would have to do a start + wait for requestNext?

Or would you keep the `ShardingProducerController` around, with a buffer and then let all requests go through that?

Perhaps for example consuming a stream and passing elements to sharding would work because then you'd have a single producer repeatedly emitting?

patriknw replied:
Yes, there should be one ShardingProducerController per node/entityType (or producer source). Messages for all entities of that type are sent via that.

For sharding the RequestNext protocol can't be as strict as for plain ProducerController because you should be able to send a new message to a new entity event though there is no demand from the already used entities. Therefore additional demand information is included in the RequestNext. It's even allowed to send to an entity that doesn't have demand and that will be buffered by the ShardingProducerController.

For the HTTP endpoint you would probably use ask via the RequestNext.askNextTo so that you get a reply when it has been confirmed.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/DurableProducerQueue.scala:
```diff
@@ -0,0 +1,41 @@
+
+  type SeqNr = Long
+
+  type ConfirmationQualifier = String
```
For both work pulling and sharding the DurableProducerQueue is shared shared for all destinations (workers/entities). Confirmation that a sequence number has been processed by the consumer must be managed per destination.

For example, 1, 2, 3 are sent to entity-1, and 4, 5 sent to entity-2. If entity-2 confirms #5 before entity-1 has sent confirmations that should mean that 4 and 5 are confirmed, but 1,2,3 are still unconfirmed.

Therefore this ConfirmationQualifier is given in the StoreMessageSent and StoreMessageConfirmed.

For sharding this corresponds to the entityId.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ShardingConsumerController.scala:
```diff
@@ -0,0 +1,80 @@
+          case other =>
+            stashBuffer.stash(other)
+            Behaviors.same
+        }
```
Yes, that should be a very common usage of stashing that we could provide some convenience for. We have a ticket for something related in https://github.com/akka/akka/issues/25824
The original attempt was going down the wrong rabbit hole though. Making more specific might make it possible.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ShardingProducerController.scala:
```diff
@@ -0,0 +1,429 @@
+              region ! ShardingEnvelope(entityId, seqMsg)
+            }
+            val p = context.spawn(ProducerController[A](outKey, durableQueueBehavior = None, send), entityId)
+            p ! ProducerController.Start(requestNextAdapter)
```
I can't see we ever GC these, in a sharded scenario isn't there a risk we end up with one PC per sharded entity after a while, since requests for any entity could come to "our" node?

patriknw replied:
yes, I'll add FIXME about cleanup of unused

patriknw replied:
Maybe we can watch the corresponding ConsumerController, to be notified of entity passivation. It should still survive a rebalance, so some idle time aspect is still needed.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ShardingProducerController.scala:
```diff
@@ -0,0 +1,429 @@
+            // no demand, buffer
+            // FIXME limit the buffers.
+            context.log.info("Buffering message to entityId [{}], buffer size [{}]", entityId, buffered.size + 1)
+            val newBuffered = buffered :+ Buffered(totalSeqNr, msg, replyTo)
```
Hmm, wouldn't it be better with per entity demand protocol of some sort, rather than a generic we can probably send to the shard so give me next and then ending up sticking it in a buffer here?

johanandren replied:
Ah, I missed out on the comment about allowing to send even when the entity isn't started so there is no demand yet.

patriknw replied:
and in the `RequestNext` there is additional demand information so a cooperative producer can use that to play nice, e.g. slow down or reject.
```
      entitiesWithDemand: Set[EntityId],
      bufferedForEntitiesWithoutDemand: Map[EntityId, Int]
```

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,416 @@
+      extends Command[A]
+  private final case object Retry extends InternalCommand
+
+  final case class Delivery[A](producerId: String, seqNr: SeqNr, msg: A, confirmTo: ActorRef[Confirmed])
```
For the confirmTo actor when used with MessageWithConfirmation how would a worker report a failure? I'm thinking of the WorkPulling scenario where a job fails, it would be nice to fail fast rather than wait for the timeout

Instead of a single class Confirmed could we include a Success/Failure?

patriknw replied:
Is that for validation errors?

What does a such error mean? Should it be resent to another worker? Is it a reply to the origin producer?

What does such error mean for plain ProducerController/ConsumerController? Rejected message should be dropped, but still confirmed.

I don't think we should build replies into this. Would become too complicated, I think. Ordinary reply can be sent to and ActorRef in the message, but that wouldn't be via reliable delivery.

chbatey replied:
Yep okay

chbatey reviewed (commented):
Still looking at this, just adding a few comments in the mean time

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,213 @@
+          // FIXME support DurableProducerQueue
+          val p =
+            context.spawnAnonymous(ProducerController[A](outKey, durableQueueBehavior = None))
+          p ! ProducerController.Start(workerRequestNextAdapter)
```
yes, PC-CC always work in pairs. Here it's one for each consumer/worker that registers its ConsumerController to the receptionist.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,508 @@
+
+}
+
+private class ProducerController[A: ClassTag](
```
why is that? everything public should be in the `object ProducerController`

chbatey replied:
I don't think you'll be able to use it in type signatures if it is private

patriknw replied:
this works fine:
```
package foo

import akka.actor.typed.Behavior
import akka.actor.typed.internal.delivery.ProducerController

object Foo {
  val start: ProducerController.Start[String] = ProducerController.Start[String](null)

  val b: Behavior[ProducerController.Command[String]] = ProducerController[String]("p1", None)

}
```

do you mean something else?

chbatey replied:
Ah i'd tried have a field be ProducerController at some point in the sample then realised it needed to be an ActorRef[ProducerController.Command[String]] or a behavior.

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,568 @@
+
+  final case class Start[A](producer: ActorRef[RequestNext[A]]) extends Command[A]
+
+  final case class RequestNext[A](sendNextTo: ActorRef[A], askNextTo: ActorRef[MessageWithConfirmation[A]])
```
For work pulling can the producer have multiple outstanding request nexts or is it still 1:1?

patriknw replied:
There it's 1:1, but when PC receives a message there might not be any worker with demand any more, since they might have been deregistered in the meanwhile. Such are stashed.

chbatey replied:
Does that mean in the WorkPulling case there's only one outstanding piece of work as there isn't a RequestNext back to the producer until it is confirmed? If i had a 1:N work pulling I'd expect at least N outstanding bits of work at any given time

patriknw replied:
You can have N*request-window outstanding work.

WorkPullingProducer will send RequestNext to the application producer as long as there is any demand from any worker. Each ProducerController for a worker will send RequestNext to the WorkPullingProducer as long as there are more space in the request window (the `upToSeqNr` in the `Request` sent from the ConsumerController).

chbatey replied:
Ah sorry i misinterpreted it being 1:1

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/DurableProducerQueue.scala:
```diff
@@ -0,0 +1,30 @@
+import akka.actor.typed.ActorRef
+
+// Implement with EventSourceBehavior in akka-persistence-typed
+// FIXME how much of this should be public? Should it be possible to plug-in different implementation?
```
yes, good to start private
if the implementation is an EventSourcedBehavior it's nice to have the API as a Behavior so it's lifecycle can be managed my the owning PC.

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,508 @@
+  final case class MessageWithConfirmation[A](message: A, replyTo: ActorRef[SeqNr]) extends InternalCommand
+
+  object Internal {
+    final case class Request(confirmedSeqNr: SeqNr, upToSeqNr: SeqNr, supportResend: Boolean, viaTimeout: Boolean)
```
would another sequence number be useful here? For the scenario that you have a fast producer and slow consumer, the consumer's buffer will full up so upTo will be say 20, then the confirmed will very slowly increase as the consumer acks the messages but in the mean time the producer will keep assuming the messages are lost and resend/extend the sequence number range. This could be avoided with a sequenceNr the consumer has stashed

patriknw replied:
PC will not push resends. The CC will request resending when it detects gaps.
The `first` message is slightly different, it will be resent by PC.

chbatey replied:
Ah i had a bug in the converted distributed workers sample that was resending the same messages over and over :))

chbatey commented:
I've mainly looked at the work pulling API. It took me a while to get my head around the various controllers but the tight integration with the receptionist means I think it is easy to use. You can see the sample here: https://github.com/akka/akka-samples/pull/186

What scenarios do we see the PC/CC being used directly? Or do we see it mainly as a building block for work pulling and sharding? Seems hard to know how you'd discover both sides to register the consumer to the producer. It could be via normal message passing or the receptionist but that's a lot of setup.

I'll take a look at sharding on Monday

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,568 @@
+import akka.util.Timeout
+import akka.actor.typed.internal.delivery.DurableProducerQueue.ConfirmationQualifier
+
+object WorkPullingProducerController {
```
Maybe FanoutProducerController?

patriknw replied:
I think Work Pulling is a rather established pattern name and the purpose is exactly for that and nothing more so I'd prefer to keep WorkPulling in the name.

patriknw commented:
I have updated the PR description with overview diagrams and summary descriptions of how Work Pulling and Sharding works. Hopefully makes it easier to understand how things are connected to each other. https://github.com/akka/akka/pull/28155#issue-338693105

patriknw commented:
@johanandren @chbatey I have completed most of this, so it's a good time for another review round.

Some remaining FIXMEs and serialization.

ismaelhamed commented:
Akka Typed only, or is this going to be implemented in classic Akka too?

patriknw commented:
It can be used from classic actors also. It’s a good question and we should add that to the docs.

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+        receiveRetry(s, () => waitingForConfirmation(retryRequest(s), seqMsg))
+
+      case start: Start[A] =>
+        start.deliverTo ! Delivery(seqMsg.producerId, seqMsg.seqNr, seqMsg.msg, context.self)
```
So the user consumer has restarted and we send the last message again

patriknw replied:
exactly

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+        }
+        Behaviors.same
+
+      case Retry =>
```
The retrying can be quite noisy if the producer is slow. Could we change it so that the timer is reset when we receive a SequencedMessage (even if it is stashed?) and perhaps make it an exponential backoff to quite a low max for when things are just quiet?

patriknw replied:
made a TODO of that for now. I agree that it would be nice to optimize that I think it has to be more sophisticated than just rescheduling for each incoming

chbatey commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala:
```diff
@@ -0,0 +1,196 @@
+
+  }
+
+  class ImageWorkManager(
```
final

chbatey commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala:
```diff
@@ -0,0 +1,196 @@
+      }
+    }
+
+    def active(next: WorkPullingProducerController.RequestNext[ImageConverter.ConversionJob]): Behavior[Command] = {
```
private

chbatey commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala:
```diff
@@ -0,0 +1,196 @@
+        case WrappedRequestNext(next) =>
+          stashBuffer.unstashAll(active(next))
+        case c: Convert =>
+          stashBuffer.stash(c)
```
deal with stash full?

chbatey commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala:
```diff
@@ -0,0 +1,196 @@
+import com.github.ghik.silencer.silent
+
+@silent("never used")
+object WorkPullingDocExample {
```
Nice example

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,215 @@
+ * the actual messages is sent to the `WorkPullingProducerController`. In that case the message is
+ * buffered and will be delivered when a new worker is registered or when there is new demand.
+ *
+ * The producer and `WorkPullingProducerController` actors are supposed to be local so that these messages are
```
And this is enforced by there being no serializer? Do you think it is worth having a way to exclude types from ever been serialized with the Java serializer?

patriknw replied:
added enforcement check by looking at ref.path.address.hasGlobalScope

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,215 @@
+import com.typesafe.config.Config
+
+/**
+ * Work pulling is a pattern where several worker actors pull tasks in their own pace from
```
Great docs, very clear

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/WorkPullingProducerController.scala:
```diff
@@ -0,0 +1,215 @@
+
+  import WorkPullingProducerControllerImpl.UnsealedInternalCommand
+
+  sealed trait Command[A] extends UnsealedInternalCommand
```
Pity to duplicated the protocol but I guess it stops accidnets when using multiple types of producer controllers

patriknw replied:
they are also slightly different

chbatey reviewed (commented):
Looking very good. I've updated the samle with the latest.

I've reviewed the interals for normal producer/consumer.

I'll look at the internals/API for work pulling/sharding next.

chbatey reviewed (commented):
Looking at the sharding API.

Do you think it would be possible to hide the fact a consumer is taking part in reliable delivery by introducting a type to indicate a message needs a reply and reliable delivery uses that to record the confirmation?

Then we could hide the SequencedMessage from the types and reduce the machinery for a consumer.

Maybe then we could provider tighter integration with sharding as I thnk it is important to allow reliable and non reliable access to the same entity.

I see a common use case of a stream of messages from say Kafka and using reliable delivery to send them in a flow controlled way to a sharded entity. This will offer a big improvement over having to tell users to do an ask from mapAsync.

Then at the same time provide read access to the entity over HTTP/gRPC. The latter messages are one off mesages that don't need to do via reliable delivery and are typically asks from a HTTP route or gRPC service implementation where it would be a lot of overhead to create a producer controller etc for a single message.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+
+                  val activeBehavior =
+                    new ConsumerControllerImpl[A](context, timers, stashBuffer, settings)
+                      .active(initialState(context, s))
```
good catch, fixed

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+          registering = s.updatedRegistering(seqMsg)),
+        seqMsg)
+    } else {
+      context.log.infoN(
```
changed to warning

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+            seqNr)
+          Behaviors.same
+        } else if (s.isNextExpected(seqMsg)) {
+          deliver(s.copy(receivedSeqNr = seqNr, registering = s.updatedRegistering(seqMsg)), seqMsg)
```
right, removed all of them after that else

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+        } else if (s.isNextExpected(seqMsg)) {
+          deliver(s.copy(receivedSeqNr = seqNr, registering = s.updatedRegistering(seqMsg)), seqMsg)
+        } else if (seqNr > expectedSeqNr) {
+          context.log.infoN("from producer [{}], missing [{}], received [{}]", pid, expectedSeqNr, seqNr)
```
changed, several

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+          throw new IllegalStateException(
+            s"Expected confirmation of seqNr [$expectedSeqNr], but received higher [$seqNr]")
+        } else if (seqNr != expectedSeqNr) {
+          context.log.info(
```
Maybe I was thinking that something could be in flight from a restart, but then it should first be reset via a new `Start`. Changed so that it's only accepting the expected.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+        stashBuffer.unstashAll(active(s.copy(confirmedSeqNr = seqNr, requestedSeqNr = newRequestedSeqNr)))
+
+      case seqMsg: SequencedMessage[A] =>
+        if (stashBuffer.isFull) {
```
don't want to complicate that at the moment

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+        val seqNr = seqMsg.seqNr
+
+        if (s.isProducerChanged(seqMsg)) {
+          receiveChangedProducer(s, seqMsg)
```
correct

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ProducerController.scala:
```diff
@@ -0,0 +1,257 @@
+   * For sending confirmation message back to the producer when the message has been confirmed.
+   * Typically used with `context.ask` from the producer.
+   *
+   * If `DurableProducerQueue` is used the confirmation reply is sent when the message has been
```
That's an interesting consideration. My thinking is that you would use a durable queue for safe (quick) handoff. Once it's stored it will eventually be handled. That ack is more interesting to use as signal of when to return OK to a HTTP request. If the full processing was quick the durable queue wouldn't be needed, since it could then be a responsibility of the original client to resend.

For the ProducerController it's possible to see the processing confirmations in the `RequestNext` message by correlating sequence numbers, but that is hidden for Sharding and WorkPulling.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ProducerControllerImpl.scala:
```diff
@@ -0,0 +1,567 @@
+      settings: ProducerController.Settings): Behavior[Command[A]] = {
+    Behaviors
+      .setup[InternalCommand] { context =>
+        context.setLoggerName("akka.actor.typed.delivery.ProducerController")
```
nah, logger name should correspond to class names. I have been thinking of adding some MDC to the logging though.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+    val newRequestedSeqNr = if (resendLost) s.requestedSeqNr else s.receivedSeqNr + flowControlWindow / 2
+    context.log.info("retry Request [{}]", newRequestedSeqNr)
+    // FIXME may watch the producer to avoid sending retry Request to dead producer
+    s.producer ! Request(s.confirmedSeqNr, newRequestedSeqNr, resendLost, viaTimeout = true)
```
ok, done

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ConsumerController.scala:
```diff
@@ -0,0 +1,229 @@
+   * application code but is public because it's possible to wrap it or send it in other ways when
+   * building higher level abstractions that are using the `ProducerController`.
+   */
+  final case class SequencedMessage[A](producerId: String, seqNr: SeqNr, msg: A, first: Boolean, ack: Boolean)(
```
I don't understand, can you clarify?

chbatey replied:
Say i have Command, A extends Command, B extends Command. I'd like a SequencedMessage[Command] to be a SequencedMessage[A]. That way if i expose the sharding reference to multiple places I can narrow the messages that area of the code can send. E.g. just the query commands to the gRPC/HTTP

patriknw commented on akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/delivery/ShardingProducerController.scala:
```diff
@@ -0,0 +1,236 @@
+  def create[A](
+      messageClass: Class[A],
+      producerId: String,
+      region: ActorRef[ShardingEnvelope[ConsumerController.SequencedMessage[A]]],
```
leaving as a TODO for now

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala:
```diff
@@ -0,0 +1,467 @@
+          context.log.info("Unstash [{}]", stashBuffer.size)
+        stashBuffer.unstashAll(active(s.copy(confirmedSeqNr = seqNr, requestedSeqNr = newRequestedSeqNr)))
+
+      case seqMsg: SequencedMessage[A] =>
```
I don't think there will be lots of duplicates since messages are not pushed from the producer, apart from the first message. Anyway, I added a dedup if the incoming is the same as we are awaiting

johanandren commented:
One round of actually-using-the-API-review, felt quite straight forward only relaying on the diagrams, feedback notes from that:

* logging at INFO from the PC and CC that should be DEBUG (or perhaps even TRACE given that it can be several log entries per message)
* Would be nice with factory methods with as few params as possible for getting started and not using durable store/changing settings.

patriknw commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/ShardingDocExample.scala:
```diff
@@ -0,0 +1,223 @@
+      val TypeKey = EntityTypeKey[ConsumerController.SequencedMessage[TodoList.Command]]("todo")
+
+      val region = ClusterSharding(system).init(Entity(TypeKey)(entityContext =>
+        ShardingConsumerController(start => TodoList(entityContext.entityId, db, start))))
```
That is `ShardingConsumerController[TodoList.Command, TodoList.Command]`. The `B` type doesn't have to be the Command type of the Behavior, i.e. in this example AddTask and Complete task doesn't have to extend TodoList.Command.

I'm pondering if that flexibility is just utterly confusing and we should only have one type? wdyt @chbatey and @johanandren ?

johanandren replied:
I'm leaning towards one type, but maybe I'm missing out on something?

patriknw commented:
> logging at INFO from the PC and CC that should be DEBUG (or perhaps even TRACE given that it can be several log entries per message)

yes, all of the logging is still on my FIXME list

patriknw commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/ShardingDocExample.scala:
```diff
@@ -0,0 +1,223 @@
+          save(newState, seqNr, confirmTo)
+          active(newState)
+        case SaveSuccess(seqNr, confirmTo) =>
+          confirmTo ! ConsumerController.Confirmed(seqNr)
```
maybe it would be much more convenient to just have a `ActorRef[Done]` for the confirmTo? Just one thing to keep around.

johanandren replied:
Hmm, yeah because it is anyway only one in flight, so the `seqNr` isn't that interesting?

johanandren replied:
Ah, you already did change this...

johanandren commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+### Work pulling delivery semantics
+
+For work pulling the order of the messages should not matter, because each message is routed randomly
+to one of the workers with demand.
```
Adding "and can therefore be processed in any order" or something along those lines would make it more clear?

johanandren commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+limited by a flow control window. The flow control is driven by the consumer side, which means that
+the `ShardingProducerController` will not send faster than the demand requested by the consumers.
+
+Example of `TodoList` entity (consumer):
```
Example subheader above would separate it better.

A short introduction of the sample would be nice to set up for faster understanding the async juggle before diving into the sample code, something like: "The sharded entity is a todo list which uses an async database call to store its entire state on each change, and first when that completes replies to reliable delivery that the message was consumed"

johanandren commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+  group=com.typesafe.akka
+  artifact=akka-actor-typed_$scala.binary_version$
+  version=$akka.version$
+}
```
For the sharding part you'd need akka-cluster-sharding-typed as well

johanandren reviewed (commented):
One round of review through the docs, looking good!

renatocaval commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+that have demand. It is allowed to send to a new `entityId` that is not included in the `RequestNext.entitiesWithDemand`.
+If sending to an entity that doesn't have demand the message will be buffered. This support for buffering
+means that it is even allowed to send several messages in response to one `RequestNext` but it's recommended to
+only send one message and wait for next `RequestNext` before sending more messages.
```
Does that mean that I can send many messages to different `entityIds` for one single `RequestNext`? What will be the impact of the confirmation in that case? 

The PC sends one request to P, P sends many messages to different entities. The entities confirm them all to the CC, then it will send many Requests back to the PC. This could cause an exponential increase in traffic, no?

renatocaval replied:
I think this breaks the flow control. If I can send more reponses then requested, then I'm sending messages to actors that didn't request it.

patriknw replied:
For sharding we can't be too strict with the demand, because it must be possible to send to new entities. Therefore it has support for buffering, but it's up to the user to use it wisely by inspecting the information in the RequestNext.

octonato reviewed (commented):
@patriknw, I did a review of the docs only. Amazing stuff.

I left a few comments and nitpicking here and there. 

I would like to review the code as well, but 16000 lines are discouraging. I need to prepare myself psychologically for the challenge. 🍷

renatocaval commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+without knowing their individual capacity and current availability.
+
+One important property is that the order of the messages should not matter, because each
+message is routed randomly to one of the workers with demand. In other words, two subsequent
```
hmm, I think I understand now what you mean here. It's not as demanded, but to workers that have demand

patriknw commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+consumer side, which means that the `WorkPullingProducerController` will not send faster than the
+demand requested by the workers.
+
+Example of image converter worker (consumer):
```
I added a `resultId: UUID` in the reply so that the client can retrieve the stored image later.

patriknw commented:
Incorporated feedback from Johan f870b49ac3c41178e7900550b5c9de5aca08ef47 and Renato 9faae8b1da0c78b28462d6c4784d887b5b0c4ad2

Fixed most of the FIXMEs. Some related to sharding were difficult but I think the solutions became alright.

patriknw commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+yet. Meaning at-least once delivery.
+
+If a worker crashes or is stopped gracefully the unconfirmed messages will be redelivered to other workers.
+In that case some of these may already have been processed by the previous worker. Meaning at-least once delivery.
```
Not sure I understand. The previous section was describing producer crashes and this section is about worker (consumer) crashes. Both may result in at-least once.

renatocaval replied:
Yes, but it still repeating the same. 

On the previous paragraph, we explain that messages may be redelivered because the confirmation is lost which give us at-least-once semantics. 

Then we kind of explain the same but in the context of crashing workers. 

The bottom line is: confirmation messages can be lost because the producer crashed before it receives it or worker crashes after processing and before confirming. In both cases, the durable queue on the producer side will ensure that the message is re-delivered to a worker (not necessarily the same as before) giving us at-least-once semantics. 

Anyway, I think we can re-write these two paragraphs by focusing on the fact that the durable queue gives at-least-once semantics and then explain the two cases in which that can happen.

patriknw replied:
I think this is just a matter of taste of how to phrase it. I wrote it based on two scenarios, a) producer crashes, b) consumer crashes. In such semantical description I prefer to repeat some things to make it clear, rather than having a nice prose. I'm leaving this as is for now, but you are welcome to open a separate PR with suggested changes if you think it's important.

patriknw commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,358 @@
+message to the `ProducerController`. Thereafter the producer will receive a new `RequestNext`
+when it's allowed to send one more message.
+
+The producer and `ProducerController` actors are supposed to be local so that these messages are
```
It's checked and enforced.

renatocaval replied:
so better if we say they are guanrateed to be local instead of supposed
```suggestion
The producer and `ProducerController` actors are guaranteed to be local so that these messages are
```

patriknw commented:
Ready for final approval.
I have completed the Java doc examples.
Created issues for remaining TODO/FIXME that can be worked on later: https://github.com/akka/akka/issues?q=is%3Aopen+is%3Aissue+label%3At%3Areliable-delivery

Let me squash this manually and craft the commit message before merging.

renatocaval commented on akka-docs/src/main/paradox/typed/reliable-delivery.md:
```diff
@@ -0,0 +1,413 @@
+while waiting for the confirmation are stashed by the `ConsumerController` and delivered when the previous
+message is confirmed.
+
+The consumer and the `ConsumerController` actors are supposed to be local so that these messages are fast
```
```suggestion
The consumer and the `ConsumerController` actors are guaranteed to be local so that these messages are fast
```

johanandren replied:
That doesn't seem right, that's up to the user, it's a recommendation for the user and not an aspect built into the reliable delivery actors.

patriknw replied:
it's enforced by a check. I'll reformulate this, but "guaranteed" is not the right word because user can do the wrong thing (but will then notice at runtime)

patriknw replied:
clarified in https://github.com/akka/akka/pull/28155/commits/64178fa5a2ff41eb59a87c8d0178317ba5ae3dc1

johanandren commented on akka-cluster-sharding-typed/src/test/java/jdocs/delivery/ShardingDocExample.java:
```diff
@@ -0,0 +1,385 @@
+                      consumerController.tell(new ConsumerController.Start<>(deliveryAdapter));
+
+                      return new Active(context, id, db, initial.state);
+                    })
```
A little bit tricky to follow with the nested lambdas and FP-and-then-OO style state, I can see that I would do the same but maybe worth doing an `Initial` `AbstractBehavior` class to make it easier for the Java crowd?

patriknw replied:
ok, changed several of them to AbstractBehavior in https://github.com/akka/akka/pull/28155/commits/dba92f361636805efe4dd8f9b40982685dface95

johanandren commented on akka-cluster-sharding-typed/src/test/java/jdocs/delivery/ShardingDocExample.java:
```diff
@@ -0,0 +1,385 @@
+    Behaviors.setup(
+        context -> {
+          // #init
+          final DB db = null; // DB implementation
```
"pretend we got it from somewhere" or something clarifying we don't actually mean it should be null

patriknw replied:
fixed in https://github.com/akka/akka/pull/28155/commits/dba92f361636805efe4dd8f9b40982685dface95

johanandren reviewed (commented):
Java samples mostly looking good, two small things noted.

johanandren reviewed (approved):
LGTM, let's merge this behemoth!

patriknw commented:
thanks, I'll squash

patriknw committed:
Reliable delivery in Typed, #20984

Different approach than in classic AtLeastOnceDelivery because I would like:
* support flow control, with a work pulling approach
* be possible to use with or without persistence (without it may loose
  messages if producer node crashes)
* detect lost messages on the consumer side and let that drive resends,
  instead of aggressively resending from producer side
* deliver messages in order and deduplicate resent messages
* have an efficient protocol for acknowledgments over the network (not ack each message),
  but still have a simple one-by-one protocol for the end user

* support 3 use cases (building blocks)
  * point-to-point
  * work pulling
  * sharding
* optional durable queue, with one event sourced implementation
* protobuf serialization
* ApiMayChange
* reference docs and examples
* api docs
* doc example code missing so far

patriknw commented:
I count @chbatey 's previous reviews as an approval

agapic commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/ShardingDocExample.scala:
```diff
@@ -0,0 +1,217 @@
+      val producerController =
+        context.spawn(ShardingProducerController(producerId, region, durableQueueBehavior = None), "producerController")
+
+      context.spawn(TodoService(producerController), "producer")
```
How do I actually get the createReceive() function to be invoked inside of TodoService? i.e., how do I actually send the UpdateTodo message?

agapic commented on akka-cluster-sharding-typed/src/test/scala/docs/delivery/ShardingDocExample.scala:
```diff
@@ -0,0 +1,217 @@
+      val producerController =
+        context.spawn(ShardingProducerController(producerId, region, durableQueueBehavior = None), "producerController")
+
+      context.spawn(TodoService(producerController), "producer")
```
Wondering if you could also include an example of actually calling UpdateTodo in the TodoService? What would that look like?

patriknw replied:
Try "todoService" in the EntityTypeKey instead, because I think the example is already using "todo" for the `TodoList` and here you are trying to interact with the `TodoService`.

However, I think you misunderstand how it is intended to be used. You try to use it with two levels of Sharding, one for the TodoList and one for the TodoService. The docs say " A single ShardingProducerController per ActorSystem (node) can be shared for sending to all entities of a certain entity type." https://doc.akka.io/docs/akka/current/typed/reliable-delivery.html#sharding

If you have more questions I'd recommend switching over to https://discuss.akka.io/
